i.MX IPM

stanislav-poboril opened the pull request:
This pull request adds IPM support for i.MX via Messaging Unit peripheral. It contains driver shim and  enables the Messaging Unit on i.MX 6SoloX SoC and Udoo Neo board. The pull request also adds the sample application, which echoes data back.
I have also written a sample Linux kernel module to be used to send the data from Linux and cooperate with the Zephyr application running on Cortex-M4. But I am not sure whether I should put it to a separate repository on Github and reference it - or I should put its source code into the Zephyr repository tree.

stanislav-poboril committed:
imx: Add IMX IPM driver for i.MX socs

Add driver for i.MX Messaging Unit peripheral which can be used for
i.MX6SoloX, i.MX7D and other i.MX socs.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
imx: Add Messaging Unit peripheral support for i.MX6SoloX soc

Add Messaging Unit B peripheral support for i.MX6SoloX soc so it can
use IMX IPM.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
imx: Enabled Messaging Unit in Udoo Neo Full board

Enabled Messaging Unit B in Cortex-M4 core of Udoo Neo Full board
so it can use IMX IPM.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
sample: Add IMX IPM sample application

Add IMX IPM sample application. It reads the data sent to the
Messaging Unit and sends them back. It has been tested on Udoo Neo Full
board with the Cortex-A9 core running Linux and sending the data
into the Messaging Unit A and the Cortex-M4 running this sample Zephyr
application and reading the data from Messaging Unit B and sending it
back for Linux side to read it from Messaging Unit A.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+Overview
+********
+A simple example that can be used with multicore i.MX SoC's containing the
```
Make this a complete sentence, and change ``SoC's`` to plural spelling:

   This simple example can be used with multicore i.MX SoCs containing a
   Messaging Unit peripheral.

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+the received data is printed to the console. When sending the data back,
+it blocks until the data are read by the remote side.
+
+The i.MX Messaging Unit peripheral consists of 4 32-bit transmit and receive
```
It reads better to spell out small numbers, so:

     ... consists of four 32-bit transmit ...

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+it blocks until the data are read by the remote side.
+
+The i.MX Messaging Unit peripheral consists of 4 32-bit transmit and receive
+registers. The sample uses the option CONFIG_IPM_IMX_MAX_DATA_SIZE_4, which
```
You can automatically create a link to the config option documentation by writing this as:

    ... uses the option :option:`CONFIG_IPM_IMX_MAX_DATA_SIZE_4`, which

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+The i.MX Messaging Unit peripheral consists of 4 32-bit transmit and receive
+registers. The sample uses the option CONFIG_IPM_IMX_MAX_DATA_SIZE_4, which
+effectively creates 4 IPM channels. Selecting the option
+CONFIG_IPM_IMX_MAX_DATA_SIZE_8 or CONFIG_IPM_IMX_MAX_DATA_SIZE_16 would result
```
same here to have the option be a link to its documentation:

    :option:`CONFIG_IPM_IMX_MAX_DATA_SIZE_4` or 
    :option:`CONFIG_IPM_IMX_MAX_DATA_SIZE_16`

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+in two 64-bit channels or a single 128-bit channel respectively.
+
+Note that this is just a simple demo to demonstrate the i.MX IPM functionality
+and blocking sending of the data back is done in the interrupt which would not
```
clarify with:

    ... is done in the interrupt handler, which would not ...

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+Note that this is just a simple demo to demonstrate the i.MX IPM functionality
+and blocking sending of the data back is done in the interrupt which would not
+be appropriate for the real world application.
```
for a real world

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+The sample requires that data are being sent by the remote core so it can echo
+it back. The other core is usually not running Zephyr but Linux on i.MX
+multicore processors, this code is not part of this Zephyr sample.
```
Start a new sentence:

    multicore processors.  This remote core code is not ...

dbkinder commented on samples/subsys/ipc/ipm_imx/README.rst:
```diff
@@ -0,0 +1,53 @@
+   :goals: build flash
+   :compact:
+
+Follow the instructions from board documentation on how to load the Zephyr
```
Change to include a reference to the board documentation:

    Follow the instructions in the :ref:`udoo_neo_full_m4` board documentation
    for how to load the Zephyr binary to the desired core and execute it.

stanislav-poboril replied:
I have changed this, thank you.

dbkinder reviewed (changes requested):
Thanks for the documentation.  Here are some recommended changes.

andrewboie commented on samples/subsys/ipc/ipm_imx/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+sample:
+    description: TBD
```
Can we fill this information in?

stanislav-poboril replied:
I have added the name and description, thank you.

tyrantlizard77 reviewed (approved):
Overall LGTM

dbkinder reviewed (approved):
LGTM, thanks!

stanislav-poboril commented:
I have a question about where to put the Linux part of the demo (which would run on the Cortex-A core).
Could it be put into the folder alongside with the Zephyr sample? But it would need to have some sort of license which is OK for both Linux and Zephyr. Would BSD-3-Clause license be OK for this Linux code (kernel module) to appear in the Zephyr repository?
Thanks

MaureenHelm commented:
The driver looks good, but can we unify the IPM samples into one sample with multiple build configurations?

stanislav-poboril commented:
> The driver looks good, but can we unify the IPM samples into one sample with multiple build configurations?

There are three IPM examples:

- ipm_mailbox (to be run on Intel Quark SE C1000)
    It uses platform specific things from ipm_quark_se.h.
- ipm_mcux (to be run on NXP LPCXpresso54114)
    It uses Zephyr on both cores, prints to serial console from master core only.
- ipm_imx (this pull request, to be run on NXP i.MX 6SoloX)
    It uses Zephyr on M4 core only and is intended to communicate with Linux on another core.
    We could unify this one with ipm_mcux, but the user would have to go into the remote subfolder and    build from there. Also having the option to print to serial console is handy this time, unlike when running Zephyr on both cores.

So there is some opportunity to unify, but it could become more clumsy for the end user and having it unified could limit adding of features which are useful in one scenario only (for example with Linux demo, we need to reply data back in a blocking fashion, so Linux can read what was written to a device - other IPM samples have different functionality).

MaureenHelm commented:
> I have a question about where to put the Linux part of the demo (which would run on the Cortex-A core).
> Could it be put into the folder alongside with the Zephyr sample? But it would need to have some sort of license which is OK for both Linux and Zephyr. Would BSD-3-Clause license be OK for this Linux code (kernel module) to appear in the Zephyr repository?

@nashif @galak @carlescufi ^^^

galak commented:
> I have a question about where to put the Linux part of the demo (which would run on the Cortex-A core).
> Could it be put into the folder alongside with the Zephyr sample? But it would need to have some sort of license which is OK for both Linux and Zephyr. Would BSD-3-Clause license be OK for this Linux code (kernel module) to appear in the Zephyr repository?
> 
> @nashif @galak @carlescufi ^^^

is this something that could live in the linux kernel itself in linux/samples/ ?

stanislav-poboril commented:
> I have a question about where to put the Linux part of the demo (which would run on the Cortex-A core).
    Could it be put into the folder alongside with the Zephyr sample? But it would need to have some sort of license which is OK for both Linux and Zephyr. Would BSD-3-Clause license be OK for this Linux code (kernel module) to appear in the Zephyr repository?

>    @nashif @galak @carlescufi ^^^

> is this something that could live in the linux kernel itself in linux/samples/ ?

I am not sure, it is related to the Zephyr sample of i.MX Messaging Unit peripheral.
I can put it into separate repository on GitHub, in case it cannot be in the Zephyr repository.

MaureenHelm commented:
The TSC agreed this week that it is ok to include the Linux part of the sample with a BSD-3 license in the Zephyr tree. The non-Apache license in the main part of the tree is acceptable in this case because it is a sample and will never be compiled into a Zephyr application. Will you please add it to the commit "sample: Add IMX IPM sample application"?

stanislav-poboril commented:
> The TSC agreed this week that it is ok to include the Linux part of the sample with a BSD-3 license in the Zephyr tree. The non-Apache license in the main part of the tree is acceptable in this case because it is a sample and will never be compiled into a Zephyr application. Will you please add it to the commit "sample: Add IMX IPM sample application"?

Yes, I am updating the documentation and will add it to the commit, but I have encountered one problem. If I use BSD license, compiling of the kernel module from example fails:

`FATAL: modpost: GPL-incompatible module mu.ko uses GPL-only symbol 'platform_driver_unregister'` 

Therefore it seems that the example kernel module needs to have a dual license:

`MODULE_LICENSE("Dual BSD/GPL");`

Does it change anything, is it still ok to include in the Zephyr source tree?

stanislav-poboril commented:
I have added the Linux kernel module code and updated the sample documentation to describe the Linux part. Note that the kernel module file samples/subsys/ipc/ipm_imx/linux/mu.c has dual BSD/GPL license otherwise it cannot build. Let me know if it is ok.

galak reviewed (changes requested):
I'm concerned about the zephyr_mu.patch.  Is this something that can just get upstream'd to linux or the UDOOboard kernel?

stanislav-poboril commented:
> I'm concerned about the zephyr_mu.patch. Is this something that can just get upstream'd to linux or the UDOOboard kernel?

@galak, are you concerned because the patch is changing the GPL-ed files or because it depends on an exact commit in the outside repository?

The patch disables the kernel's usage of Messaging Unit and replaces it with the functionality needed by the demo application. Upstreaming to UDOObuntu kernel goes against their usage of Messaging Unit. Maybe the Zephyr example's Linux side should go into a different repository at the end. Or we could fork UDOObuntu kernel into a repository controlled by us (to avoid dependency on a single commit in a third party controlled repository) and commit our changes there.

galak commented:
> The patch disables the kernel's usage of Messaging Unit and replaces it with the functionality needed by the demo application. Upstreaming to UDOObuntu kernel goes against their usage of Messaging Unit. Maybe the Zephyr example's Linux side should go into a different repository at the end. Or we could fork UDOObuntu kernel into a repository controlled by us (to avoid dependency on a single commit in a third party controlled repository) and commit our changes there.

How is UDOObuntu using the message unit?  Trying to understand how this differs from what they are doing?  Do you have a pointer to the UDOObuntu kernel repo?

stanislav-poboril commented:
@galak, UDOObuntu version of kernel is here: https://github.com/UDOOboard/linux_kernel
There is the file https://github.com/UDOOboard/linux_kernel/blob/3.14-1.0.x-udoo/arch/arm/mach-imx/mu.c which handles Messaging Unit device. It is part of the kernel image, not a loadable module, so it cannot be easily unloaded. UDOObuntu uses M4 core in an Arduino-like fashion and uses MCC + MU to communicate between A9 and M4. I wanted to bypass this communication since there will be no "Arduino-like" code on M4 but Zephyr and wanted to demonstrate the basic MU functionality without the logic already present in the existing mu.c.

galak commented:
> @galak, UDOObuntu version of kernel is here: https://github.com/UDOOboard/linux_kernel
> There is the file https://github.com/UDOOboard/linux_kernel/blob/3.14-1.0.x-udoo/arch/arm/mach-imx/mu.c which handles Messaging Unit device. It is part of the kernel image, not a loadable module, so it cannot be easily unloaded. UDOObuntu uses M4 core in an Arduino-like fashion and uses MCC + MU to communicate between A9 and M4. I wanted to bypass this communication since there will be no "Arduino-like" code on M4 but Zephyr and wanted to demonstrate the basic MU functionality without the logic already present in the existing mu.c.

I understand, we should focus on what upstream Linux supports or see what we can get supported there and not worry about the UDOOboard kernel being pretty old.

I see code for remoteproc on imx in the upstream kernel, not sure if there is a message unit driver or not in the upstream/mainline kernel.

stanislav-poboril commented:
The non-Apache-2.0 / Linux code has been moved to https://source.codeaurora.org/external/imxsupport/imx-mu/. I think the pull request could be merged now.

diegosueiro commented:
@galak ,

Any idea when this is going to be merged?

galak commented on dts/arm/nxp/nxp_imx6sx_m4.dtsi:
```diff
@@ -241,6 +241,18 @@
 			status = "disabled";
 		};
+
+		mub:mu@4229C000 {
```
Can you make this 4229C000 -> 4229c000 (and the reg as well)

stanislav-poboril replied:
Done.

galak commented on dts/bindings/arm/nxp,imx-mu.yaml:
```diff
@@ -0,0 +1,45 @@
+#
+---
+title: IMX MESSAGING UNIT
+id: nxp,imx-mu
```
remove id:

stanislav-poboril replied:
Done.

galak commented on samples/subsys/ipc/ipm_imx/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+tests:
+    test:
+        build_only: true
+        filter: CONFIG_SOC_IMX
```
This should be filter: CONFIG_SOC_FAMILY_IMX

If we do that, do we need the platform_whitelist as well?

stanislav-poboril replied:
Done - filter changed. The platform_whitelist is still needed, once @diegosueiro merges MU support into his work on i.MX7, platform_whitelist could be perhaps removed then.

galak commented on soc/arm/nxp_imx/mcimx6x_m4/dts_fixup.h:
```diff
@@ -72,27 +72,32 @@
 #define CONFIG_UART_IMX_UART_3_BASE_ADDRESS	NXP_IMX_UART_421EC000_BASE_ADDRESS
 #define CONFIG_UART_IMX_UART_3_NAME		NXP_IMX_UART_421EC000_LABEL
 #define CONFIG_UART_IMX_UART_3_IRQ_NUM		NXP_IMX_UART_421EC000_IRQ_0
-#define CONFIG_UART_IMX_UART_3_IRQ_PRI		NXP_IMX_UART_421EC000_IRQ_0_PRIORITY
```
Fix up for CONFIG_ -> DT_ change.

stanislav-poboril replied:
Done.

galak reviewed (changes requested):
Inline comments:

Also (related to inline comments), these build warnings:
udoo_neo_full_m4.dts_compiled: Warning (simple_bus_reg): /soc/mu@4229C000: simple-bus unit address format error, expected "4229c000"
extract_dts_includes.py: WARNING: id field set in 'IMX MESSAGING UNIT', should be removed.
